prov/verbs, rxm: fix for a deadlock, refactoring & add env variable to force auto-progress

a-ilango opened the pull request.

shefty commented on prov/rxm/src/rxm_init.c:
```diff
@@ -204,10 +204,32 @@ static int rxm_init_info(void)
+			info->domain_attr->data_progress =
+				hints->domain_attr->data_progress;
+	}
+}
```
Can you provide details on why this is needed in the commit message?  I.e. provides a work-around for apps that don't drive progress correctly...

a-ilango replied:
I've modified the commit to force auto-progress even if app has requested manual progress. Let me know if the one-line commit message is sufficient.

shefty commented on prov/rxm/src/rxm_init.c:
```diff
@@ -412,6 +429,11 @@ RXM_INI
+	fi_param_define(&rxm_prov, "data_progress_auto", FI_PARAM_BOOL,
+			"Set default data progress mode to FI_PROGRESS_AUTO "
+			"(default: 0). Application hints takes precedence "
+			"over this.\n");
```
Should the application hints take precedence, or do we want to force auto progress under the covers?

a-ilango replied:
I think this env variable should change only the default progress mode (when app doesn't care). If an app explicitly requires a progress mode, we can't return a different mode. Middleware like MPI just pick whatever progress mode the provider reports. So this env variable can help test different progress modes.

Although in mercury RPC's case, this variable would not be useful as their app doesn't set the progress to FI_PROGRESS_UNSPEC if I remember correctly.

shefty replied:
I don't know that changing the default return value is that useful.  We can test different modes by controlling the value in the test app.  For the Mercury problem, I think we need to enable auto progress, separate from what we actually report. That's what I thought your intent was.

a-ilango replied:
Does the API allow a provider to return auto progress when hints specify manual progress?

Changing the test app is not easy unless the app provides an env variable / command line option to do that.

I was hoping this might help mercury but on double-checking mercury actually specifies a progress mode explicitly in hints. So mercury app developers may need to provide a way to change it.

a-ilango commented:
@shefty I've updated the PR with some additional commits. The commit which removed "saving msg_ep" made it easy to expose a deadlock in the verbs provider in multi-threaded case (e.g. auto-progress). I've added a fix for that.

shefty commented on prov/rxm/src/rxm_init.c:
```diff
@@ -412,6 +414,10 @@ RXM_INI
 
+	fi_param_define(&rxm_prov, "data_progress_auto", FI_PARAM_BOOL,
+			"Force data auto-progress even if app requested manual "
+			"progress (default: 0) \n");
```
default: false or no?

a-ilango replied:
Default is false/0/no, that is auto-progress is not forced.

shefty replied:
I wasn't clear.  I was referring to the wording in the message to state the default in terms of "false" or "no" instead of an integer 0.

a-ilango replied:
Got it, I shall update that.

a-ilango commented:
Removed the commit that deleted unnecessary code in RxM connection path as it needs more fixes first. They will come in a separate PR.

@swelch I've removed the EQ lock acquire in the XRC functions that was causing issue.

swelch commented:
> 
> @swelch I've removed the EQ lock acquire in the XRC functions that was causing issue.

The patch verbs_eq.c to remove now unneeded locking looks good. I pulled the change and verified XRC was working. Thanks.

swelch commented:
@a-ilango - I have another PR coming, that probably should rebased on this one. If you think this will hit master fairly soon I will wait and add it then. Any idea when this PR will be merged (I see you removed the do not merge label)?

a-ilango commented:
@swelch Thanks for verifying. Yes, this PR is ready to be merged once the CI completes. I've removed extra patches which needed more work.
